src/componentProvider.js

```diff
@@ -22,42 +22,48 @@ export default function componentProvider(config) {
   // The component selector.
   const selector = `[data-component='${name}']`;
 
-  // Get component elements.
-  const componentEls = document.querySelectorAll(selector);
-
-  // Do nothing.
-  if (componentEls.length < 1) {
-    console.log(`No elements found for ${selector}`); // eslint-disable-line no-console
-    return undefined;
-  }
-
   // Get options.
   const options = config.options || {};
 
-  /*
-   * Collect component arguments based on the config.
+  /**
+   * The provider function.
    *
-   * @returns {Array}
+   * Collects component elements and passes them to each instance of the component.
    */
-  const componentArgs = Array.from(componentEls).map((element) => {
-    const children = {};
+  const init = () => {
+    const componentEls = document.querySelectorAll(selector);
 
-    // Select single child nodes.
-    Object.keys(querySelector).forEach((elementKey) => {
-      children[elementKey] = element.querySelector(querySelector[elementKey]);
-    });
+    if (componentEls.length < 1) {
+      // Do nothing.
+      console.log(`No elements found for ${selector}`); // eslint-disable-line no-console
+      return undefined;
+    }
 
-    // Select groups of child nodes.
-    Object.keys(querySelectorAll).forEach((elementKey) => {
-      const nodeList = element.querySelectorAll(querySelectorAll[elementKey]);
-      children[elementKey] = Array.from(nodeList);
-    });
+    /*
+     * Collect component arguments based on the config.
+     *
+     * @returns {Array}
+     */
+    const componentArgs = Array.from(componentEls).map((element) => {
+      const children = {};
+
+      // Select single child nodes.
+      Object.keys(querySelector).forEach((elementKey) => {
+        children[elementKey] = element.querySelector(querySelector[elementKey]);
+      });
 
-    return ({ element, children, options });
-  });
+      // Select groups of child nodes.
+      Object.keys(querySelectorAll).forEach((elementKey) => {
+        const nodeList = element.querySelectorAll(querySelectorAll[elementKey]);
+        children[elementKey] = Array.from(nodeList);
+      });
+
+      return ({ element, children, options });
+    });
 
-  // Create the provider function.
-  const init = () => componentArgs.forEach((args) => new Component(args));
+    // eslint-disable-next-line no-void
+    return void componentArgs.forEach((args) => new Component(args));
```

Comment:
Some food for thought: I think this would be more readable if it were split into two lines, the first the `forEach` and the second... I guess maybe the `return` statement would be unnecessary? It also might be more _useful_ if it instead returned an array of components (`return componentArgs.map(...)`) -- useful e.g. in unit tests.

Reply:
Thanks, Matt. I split the DOM querying into its own function to avoid those confusing return statements and to hopefully clarify what `componentProvider` does/returns